tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py

```diff
@@ -0,0 +1,181 @@
+"""
+Unit test for GaussianMLPPolicy.
+
+This test consists of four different GaussianMLPPolicy: P1, P2, P3
+and P4. All four policies are implemented with GaussianMLPModel.
+P1 and P2 are from GaussianMLPPolicyWithModel, which uses
+garage.tf.distributions while P3 and P4 from GaussianMLPPolicyWithModel2,
+which uses tf.distributions.
+
+It does not aim to show how GaussianMLPModel will be used in GaussianMLPPolicy,
+which is self-explanatory in the implementation of GaussianMLPPolicyWithModel.
+It aims to illustrate the differences in operations between
+GaussianMLPPolicyWithModel and GaussianMLPPolicyWithModel2.
+
+It covers dist_info_sym, kl_sym, log_likelihood_sym, entropy_sym and
+likelihood_ratio_sym. Also, it shows how a GaussianMLPPolicy should be pickled.
+"""
+import pickle
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from garage.tf.envs import TfEnv
+from garage.tf.misc import tensor_utils
+from garage.tf.policies import GaussianMLPPolicyWithModel
+from garage.tf.policies import GaussianMLPPolicyWithModel2
+from tests.fixtures import TfGraphTestCase
+from tests.fixtures.envs.dummy import DummyBoxEnv
+
+
+class TestGaussianMLPPolicyWithModel(TfGraphTestCase):
+    def test_gaussian_mlp_policy_with_model(self):
+        box_env = TfEnv(DummyBoxEnv())
+        policy1 = GaussianMLPPolicyWithModel(
+            env_spec=box_env, init_std=1.1, name='P1')
+        policy2 = GaussianMLPPolicyWithModel(
+            env_spec=box_env, init_std=1.2, name='P2')
+        policy3 = GaussianMLPPolicyWithModel2(
+            env_spec=box_env, init_std=1.1, name='P3')
+        policy4 = GaussianMLPPolicyWithModel2(
+            env_spec=box_env, init_std=1.2, name='P4')
+
+        # self.sess.run(tf.global_variables_initializer())
+
+        obs = [box_env.reset()]
+
+        obs_ph = tf.placeholder(tf.float32, shape=(None, 1))
+
+        ############################################################
+        # dist_info_sym
+        # API unchanged
+        #
+        # * Existing way *
+        # dist1_sym and dist2_sym are still dict(mean=mean, log_std=log_std)
+        dist1_sym = policy1.dist_info_sym(obs_ph, name='p1_sym')
+        dist2_sym = policy2.dist_info_sym(obs_ph, name='p2_sym')
+
+        # * New way *
+        # this will be tf.distributions
+        policy3.dist_info_sym(obs_ph, name='p3_sym')
```

Comment:
does this have a return? where is the return used?
will this func be used in GaussianMLPPolicyWithModel2? if it is not, we can discard this interface.

Reply:
Yes, this will be used. This will actually call `model.build()` with the given placeholder. We can rename it to be more clear, e.g. `policy3.build_dist(obs_ph, name="p3_sym")`.